src/datasets/arrow_dataset.py

```diff
@@ -1246,43 +1258,108 @@ def save_to_disk(self, dataset_path: str, fs=None):
+                "job_id": shard_idx,
+                "shard": dataset.shard(num_shards=num_shards, index=shard_idx, contiguous=True),
+                "fpath": path_join(dataset_path, f"data-{shard_idx:05d}-of-{num_shards:05d}.arrow"),
+                "storage_options": storage_options,
+            }
+            for shard_idx in range(num_shards)
+        )
+        shard_lengths = [None] * num_shards
+        shard_sizes = [None] * num_shards
+        if num_proc > 1:
+            with Pool(num_proc) as pool:
+                for job_id, done, content in iflatmap_unordered(pool, Dataset._save_to_disk_single, args_per_job):
+                    if done:
+                        shards_done += 1
+                        pbar.set_description(f"Saving dataset ({shards_done}/{num_shards} shards)")
+                        logger.debug(f"Finished writing shard number {job_id} of {num_shards}.")
+                        shard_lengths[job_id], shard_sizes[job_id] = content
+                    else:
+                        pbar.update(content)
+        else:
+            for args in args_per_job:
+                for job_id, done, content in Dataset._save_to_disk_single(args):
+                    if done:
+                        shards_done += 1
+                        pbar.set_description(f"Saving dataset ({shards_done}/{num_shards} shards)")
+                        logger.debug(f"Finished writing shard number {job_id} of {num_shards}.")
+                        shard_lengths[job_id], shard_sizes[job_id] = content
+                    else:
+                        pbar.update(content)
+        with fs.open(path_join(dataset_path, config.DATASET_STATE_JSON_FILENAME), "w", encoding="utf-8") as state_file:
             json.dump(state, state_file, indent=2, sort_keys=True)
         with fs.open(
-            Path(dataset_path, config.DATASET_INFO_FILENAME).as_posix(), "w", encoding="utf-8"
+            path_join(dataset_path, config.DATASET_INFO_FILENAME), "w", encoding="utf-8"
         ) as dataset_info_file:
             # Sort only the first level of keys, or we might shuffle fields of nested features if we use sort_keys=True
             sorted_keys_dataset_info = {key: dataset_info[key] for key in sorted(dataset_info)}
             json.dump(sorted_keys_dataset_info, dataset_info_file, indent=2)
-        logger.info(f"Dataset saved in {dataset_path}")
+
+    @staticmethod
+    def _save_to_disk_single(arg):
+        job_id: Dataset = arg["job_id"]
+        shard: Dataset = arg["shard"]
+        fpath: str = arg["fpath"]
+        storage_options: Optional[dict] = arg["storage_options"]
+        refresh_rate = 0.05  # 20 progress updates per sec
+        batch_size = config.DEFAULT_MAX_BATCH_SIZE
+
+        if shard._indices is not None:
+            raise ValueError(
+                "`_save_to_disk_single` only support shards with flattened indices. "
+                "Please call ds.flatten_indices() before saving to disk."
+            )
```

Comment:
`save_to_disk` calls `flatten_indices()` before calling `_save_to_disk_single`. So in practice this error never happens.

I just added it in case someone plays with the internals of the lib